### PR TITLE
Distributed OOMptimizer for speechlm2 training

### PR DESCRIPTION
**Pull request by pzelasko**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Distributed OOMptimizer extends the original OOMptimizer to support batch size estimation for models leveraging model parallelism (multi-node supported) - specifically SALMAutomodel for now. It uses a supervisor process to recover from distributed OOM/NCCL hangs. 

**Collection**: speechlm2

# Changelog

- Distributed OOMptimizer

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by pzelasko** [comment]
/ok to test [a85da02](https://github.com/NVIDIA-NeMo/NeMo/pull/15807/commits/a85da02586704dc99a5d22fb5797ac84f87d6ff9)

**Comment by pzelasko** [comment]
/ok to test [35e3553](https://github.com/NVIDIA-NeMo/NeMo/pull/15807/commits/35e3553db435272b7c686acfc08dde3ef8a0105a)

File: scripts/speechlm2/oomptimizer.py [hunk 1]
```diff
@@ -423,9 +426,7 @@ def oomptimizer(
     def get_max_seq_lens(buckets):
 
         def _determine_lens_for_bucket(bin):
```
Comment (KunalDhawan) [inline scripts/speechlm2/oomptimizer.py]: Removing the SALM special-case here changes behavior silently rather than removing support. A 1D bucket for a SALM model now hits the `audio→text` case, so `audios` becomes `compute_num_samples(bin)` seconds of audio and `input_ids` becomes `ceil(ratio*bin)` tokens (`ratio` defaults to `12`), a different definition of "bucket" than `distributed_oomptimizer.py`'s token-count semantics, and the old code treated it as `(bin, bin)`.

To address this, imo we could either:
- add an explicit guard that errors out when a SALM-family model is passed to the single-process tool, **or**
- reuse `SequenceLengthResolver` so both tools agree?
What do you think?

File: scripts/speechlm2/distributed_oomptimizer.py [hunk 2]
```diff
@@ -0,0 +1,1700 @@
...
+@click.option(
+    "--salm-audio-token-ratio",
+    type=float,
+    default=0.75,
+    help="For SALM-style 1D token buckets, fraction of the bucket represented by audio-equivalent tokens.",
+)
+@click.option(
+    "--distributed-timeout-seconds",
+    type=float,
+    default=15.0,
```
Comment (KunalDhawan) [inline scripts/speechlm2/distributed_oomptimizer.py:1305]: `15s` as the process-group timeout feels tight for large LLM + FSDP init, where the first collective can take longer than that on a cold start. A spurious timeout here gets misclassified as OOM.
Could we bump the default, or use a longer timeout for init and only tighten it for the profiling step itself?

File: scripts/speechlm2/distributed_oomptimizer.py [hunk 3]
```diff
@@ -0,0 +1,1700 @@
...
+    def wait(self, name: str) -> None:
+        if self.nnodes <= 1:
+            return
+        barrier_dir = self.log_dir / ".supervisor_barriers" / name
+        barrier_dir.mkdir(parents=True, exist_ok=True)
+        marker = barrier_dir / f"rank_{self.node_rank}.ready"
+        marker.write_text(str(os.getpid()))
+        deadline = time.monotonic() + self.timeout_seconds
+        while time.monotonic() < deadline:
+            if len(list(barrier_dir.glob("rank_*.ready"))) >= self.nnodes:
```
Comment (KunalDhawan) [inline scripts/speechlm2/distributed_oomptimizer.py:426]: `prepare()` cleans result/outcome/log files but not `.supervisor_barriers/`, so on a re-run in the same `log_dir` stale `rank_*.ready` markers from a prior run can satisfy this condition immediately and the barrier passes without actually synchronizing across nodes.

Could we namespace the barrier dir per run (e.g. include the rdzv/job id) and/or clear it in `prepare()`?

File: scripts/speechlm2/distributed_oomptimizer.py [hunk 4]
```diff
@@ -0,0 +1,1700 @@
...
+                seq_length = select_seq_length[item["seq_length"]]
+                tnsr = torch.ones(B, dtype=torch.long, device=self.device) * seq_length
+            elif isinstance(nt.elements_type, MaskType):
+                seq_length = select_seq_length[item["seq_length"]]
+                tnsr = torch.ones(B, seq_length, device=self.device, dtype=torch.bool)
+            elif isinstance(nt.elements_type, LabelsType):
+                seq_length = select_seq_length[item["seq_length"]]
+                tnsr = torch.randint(0, item["vocab_size"], size=(B, seq_length), device=self.device)
+                for token_id in item.get("excluded_token_ids", []):
+                    tnsr.masked_fill_(tnsr == token_id, 0)
```
Comment (KunalDhawan) [inline scripts/speechlm2/distributed_oomptimizer.py]: Masking excluded tokens to the literal id `0` assumes `0` is a safe filler, but it may be a meaningful/special token for some tokenizers. Should we mask to the model's pad id instead?

**Review by KunalDhawan — COMMENTED** [review]
Thanks @pzelasko, amazing effort! This is a very useful adidition for efficient distributed training. Did a first pass review and added a few questions/comments, other than that LGTM!

File: scripts/speechlm2/oomptimizer.py (hunk 1 above)
Comment (pzelasko) [reply scripts/speechlm2/oomptimizer.py]: Good catch it should support SALM since FSDP/TP works there too.

File: scripts/speechlm2/distributed_oomptimizer.py (hunk 2 above)
Comment (pzelasko) [reply scripts/speechlm2/distributed_oomptimizer.py:1305]: I didn't observe issues in practice even on a slow cluster but sure, let's add a larger initial timeout to be safe

**Comment by pzelasko** [comment]
/ok to test [a7d4a4f](https://github.com/NVIDIA-NeMo/NeMo/pull/15807/commits/a7d4a4f42c1b7d55959948c19aca9778a45b5411)

**Comment by pzelasko** [comment]
/ok to test [2381c7b](https://github.com/NVIDIA-NeMo/NeMo/pull/15807/commits/2381c7b1e980947de295b7f5a9b49d587d0ba5a4)

**Review by KunalDhawan — APPROVED** [review]
LGTM, great work @pzelasko!

**Comment by pzelasko** [comment]
I'm merging since all tests passed before merging latest main into this PR's branch.
